services/org/org.go

```diff
@@ -0,0 +1,57 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package org
+
+import (
+	"fmt"
+
+	"code.gitea.io/gitea/models"
+	"code.gitea.io/gitea/models/db"
+	"code.gitea.io/gitea/modules/storage"
+	"code.gitea.io/gitea/modules/util"
+)
+
+// DeleteOrganization completely and permanently deletes everything of organization.
+func DeleteOrganization(org *models.User) error {
+	if !org.IsOrganization() {
+		return fmt.Errorf("%s is a user not an organization", org.Name)
+	}
+
+	ctx, commiter, err := db.TxContext()
+	if err != nil {
+		return err
+	}
+	defer commiter.Close()
+
+	// Check ownership of repository.
+	count, err := models.GetRepositoryCount(ctx, org)
+	if err != nil {
+		return fmt.Errorf("GetRepositoryCount: %v", err)
+	} else if count > 0 {
+		return models.ErrUserOwnRepos{UID: org.ID}
+	}
+
+	if err = models.DeleteOrganization(ctx, org); err != nil {
+		return fmt.Errorf("DeleteOrganization: %v", err)
+	}
+
+	// FIXME: system notice
+	// Note: There are something just cannot be roll back,
+	//	so just keep error logs of those operations.
+	path := models.UserPath(org.Name)
+
+	if err := util.RemoveAll(path); err != nil {
+		return fmt.Errorf("Failed to RemoveAll %s: %v", path, err)
+	}
+
+	if len(org.Avatar) > 0 {
+		avatarPath := org.CustomAvatarRelativePath()
+		if err := storage.Avatars.Delete(avatarPath); err != nil {
+			return fmt.Errorf("Failed to remove %s: %v", avatarPath, err)
+		}
+	}
+
+	return commiter.Commit()
```

Comment:
We should commit before remove the files. because once commit failure, then database will be rollback, but files has been removed. That's broken.

Reply:
I made it like in the original code but will change it.

Reply:
Hm, so the original codes have problem.